modules/loggers/mod_log_config.c

```diff
@@ -1602,13 +1634,204 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
+        // this is an actual error, probably wrong idx
+        // idx did point to a byte which looks like an utf-8 1 byte continuation
+        return -1;
+    } else if(c < 0x80) { // 1 byte utf-8
+        code_point = c;
+        expect_more_bytes = 0;
+        len++;
+    }
+
+    while(expect_more_bytes > 0) {
+        if(++idx >= sl) break;
+        c = string[idx];
+        if((c & 0b10000000) == 0b10000000) { // 1 byte utf-8 continuation
+            c = c & 0b00111111;
+            code_point |= c;
+        } else {
+            // unexpected byte sequence, error
+            return -1;
+        }
+
+        expect_more_bytes--;
+        len++;
+
+        if(expect_more_bytes > 0) {
+            code_point <<= 6;
+        }
+    }
+
+    *cp = code_point;
+    return len;
+}
+
+/* see https://www.rfc-editor.org/rfc/rfc8259#section-7 */
+static int json_needs_encoding(const char* string)
+{
+    apr_uint32_t code_point;
+    int idx = 0;
+    int len = utf8_next_code_point(string, idx, &code_point);
+
+    while(len > 0) {
+        if(code_point  < 0x20 ||
+           code_point == 0x22 ||
+           code_point == 0x5c) {
+            return 1; // true, this json utf-8 string, needs encoding
+        }
+
+        idx += len;
+        len = utf8_next_code_point(string, idx, &code_point);
+    }
+
+    return 0;
+}
+
+static const char* json_encode(apr_pool_t *p, const char* utf8_string_to_encode)
+{
+    apr_uint32_t code_point;
+
+    struct iovec strcats[strlen(utf8_string_to_encode)];
+    apr_size_t isc = 0;
+
+    int idx = 0;
+    int len = utf8_next_code_point(utf8_string_to_encode, idx, &code_point);
+
+    while(len > 0) {
+        // unescaped = %x20-21 / %x23-5B / %x5D-10FFFF
+        if(code_point  < 0x20 ||
+           code_point == 0x22 ||
+           code_point == 0x5c) {
+            strcats[isc].iov_base = apr_psprintf(p, "\\u%04x", code_point);
+            strcats[isc].iov_len = strlen(strcats[isc].iov_base);
+        } else {
+            strcats[isc].iov_base = (void*) utf8_string_to_encode + idx;
+            strcats[isc].iov_len = len;
+        }
+
+        isc++;
+        idx += len;
+        len = utf8_next_code_point(utf8_string_to_encode, idx, &code_point);
+    }
+
+    return apr_pstrcatv(p, strcats, isc, NULL);
+}
+
+static void add_iovec_str(struct iovec *iv, const char *str)
+{
+    iv->iov_base = (void*) str;
+    iv->iov_len = strlen(str);
+}
+
+static apr_status_t ap_json_log_writer( request_rec *r,
                            void *handle,
                            const char **strs,
                            int *strl,
                            int nelts,
+                           void *itms,
                            apr_size_t len)
 
+{
+    log_format_item *items = (log_format_item *) itms;
+    apr_size_t len_file_write;
+
+    const char* attribute_name;
+    const char* attribute_value;
+    const char* json_str;
+
+    struct iovec strcats[1 + (nelts * 7) ];
```

Comment:
VLA (Variable Length Array) should be avoided in httpd, you could use an `apr_array_header_t` of `struct iovec` and modify the add_iovec_str() to push there instead.